app/models/concerns/network_analysis.rb

```diff
@@ -0,0 +1,103 @@
+module NetworkAnalysis
+
+  def interlocks(page = 1)
+    nodes_connected_by_common_neighbor(:interlocks, page)
+  end
+
+  def similar_donors(page = 1)
+    nodes_connected_by_common_neighbor(:giving, page)
+  end
+
+  private
+
+  def nodes_connected_by_common_neighbor(connection_type, page = 1)
+    id_hashes = paginate(page, Entity::PER_PAGE, connected_id_hashes_for(connection_type))
+    entities_by_id = Entity.lookup_table_for(collapse(id_hashes))
+    id_hashes.map do |id_hash|
+      {
+        "connected_entity" => entities_by_id.fetch(id_hash[:connected_id]),
+        "connecting_entities" => id_hash[:connecting_ids].map { |id| entities_by_id.fetch(id) }
+      }
+    end
+  end
+
+  # TODO (ag|Tue 03 Oct 2017): extract object for this?
+  # ---
+  # type ConnectedIdHash = { connected_id   => Integer,
+  #                          connecting_ids => [Integer] }
+  # ---
+  # [Integer], Integer -> [ConnectedIdHash]
+  def connected_id_hashes_for(connection_type)
```

Comment:
this is _very_ concise and beautifully written, but also somewhat inflexible for doing custom advanced queries.

there are lots of interlock-like queries we could want to explore. It would be nice if we could pass in a set of connecting_ids and a set of category_ids.  Perhaps we want to try creating interlocks for a number of entities combined (like from a tag or list) and a custom set of categories or a set of custom connecting entities. Having those options exposed in a public method isn't strictly necessary for the Giving and Interlocks page, so in some ways they make the code more confusing, but I think they would be useful to have for future research. Especially if we want to recreate this feature: https://littlesis.org/org/1/Walmart/networkSearch

we can address this in a future refactor since it doesn't really impact the giving and interlocks pages.

Reply:
let's do it! totally agree it would be nice to expose a custom function like this. 

maybe fold that in with extracting a `NodesWithCommonNeighbors` class and thinking carefully about what its initialization params are?

Reply:
here is a sketch of what such a class might look like. going to defer implementing it until after completing #336:

```ruby
  class NodesConnectedByCommonNeighbor
    def initialize(args)
      @root_entity = args[:root_entity] # we can infer `primary_ext` from this...
      @connection_type = args[:connection_type]
      @connecting_ids = args[:connecting_ids] || connecting_ids_for(@connection_type)
      @connecting_entity_id_key = connecting_entity_id_key_for(@connection_type)
      @relationship_categories = relationship_categories_for(@connection_type)
    end

    def self.entity_hashes_from(id_hashes)
      id_hashes.map do |id_hash|
        {
          "connected_entity" => entities_by_id.fetch(id_hash[:connected_id]),
          "connecting_entities" => id_hash[:connecting_ids].map { |id| entities_by_id.fetch(id) }
        }
      end
    end
    
    def id_hashes
      #logic from old #connected_id_hashes_for
    end

    def paginated_id_hashes
      # paginate the above
    end
  end
```